### PR TITLE
LinearSolve interface

### PR DESCRIPTION
**Pull request by vpuri3**
TODO

- [ ] interoperability `solve(prob::ODEProblem,Rodas5(linsolve=alg::SciMLLinearSolveAlgorithm))`
- [ ] wrap and test a KSP libraries - `Krylov.jl`, `IterativeSolvers.jl`, `KrylovKit.jl`. handle in-place, and out-of-place algs separately? save KSP solver cache in `LinearCache.cacheval`
- [ ] add present solution, iteration count, other functionality to `LinearCache`

ref https://github.com/SciML/LinearSolve.jl/issues/3

### CODE REVIEW

**Comment by vpuri3** [comment]
@ChrisRackauckas how to handle this?

```julia
    prob = ODEProblemLibrary.prob_ode_linear
    sol  = solve(prob, Rodas5(linsolve=SVDFactorization()); saveat=0.1)
```

```julia
  MethodError: no method matching copy(::SciMLBase.UDerivativeWrapper{ODEFunction
{false, DiffEqProblemLibrary.ODEProblemLibrary.var"#1#2", UniformScaling{Bool}, D
iffEqProblemLibrary.ODEProblemLibrary.var"#3#4", Nothing, Nothing, Nothing, Nothi
ng, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, typeof(SciMLBa
se.DEFAULT_OBSERVED), Nothing}, Float64, Float64})                               
  Closest candidates are:                                                        
    copy(::DataStructures.MutableLinkedList{T}) where T at /Users/vp/.julia/packa
ges/DataStructures/nBjdy/src/mutable_list.jl:110                                 
    copy(::BandedMatrices.BandedLU{T, S}) where {T, S} at /Users/vp/.julia/packag
es/BandedMatrices/2lafE/src/banded/BandedLU.jl:48                                
    copy(::FillArrays.Zeros) at /Users/vp/.julia/packages/FillArrays/VLeUk/src/Fi
llArrays.jl:280                                                                  
    ...                                                                          
  Stacktrace:                                                                        [1] init(::LinearProblem{DataType, false, SciMLBase.UDerivativeWrapper{ODEFun
ction{false, DiffEqProblemLibrary.ODEProblemLibrary.var"#1#2", UniformScaling{Boo
l}, DiffEqProblemLibrary.ODEProblemLibrary.var"#3#4", Nothing, Nothing, Nothing, 
Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, typeof(Sc
iMLBase.DEFAULT_OBSERVED), Nothing}, Float64, Float64}, Float64, SciMLBase.NullPa
rameters, Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}}, ::SVDFa
ctorization{LinearAlgebra.DivideAndConquer}; alias_A::Bool, alias_b::Bool, kwargs
::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})                 
      @ LinearSolve ~/.julia/dev/LinearSolve/src/common.jl:62
```

**Comment by ChrisRackauckas** [comment]
The ODE solvers would need to be updated to use LinearSolve.jl before you can pass a LinearSolve.jl solver to it.

**Comment by vpuri3** [comment]
how to go about this? or is it a later step and i should write more wrappers first?

**Comment by ChrisRackauckas** [comment]
There's a lot of random changes here. Please make the PR towards a concise point, and split to multiple PRs if necessary.

**Comment by vpuri3** [comment]
yeah i made changes at  a few different places. will break it into pieces.

**Comment by vpuri3** [comment]
Here's what I have in this branch

1. Common interface: in addition to calling `solve(problem::LinearProblem, alg)`, you can also call `alg(x, A, b)` (or `alg(prob::LinearProblem)`) where `alg isa SciMLLinearSolveAlgorithm`.  Both these calls create a new `LinearCache`. If you want to reuse a `LinearCache`, then you can create one like `cache = SciMLBase.init(prob,alg)`, and then reuse like `y=cache(x,A,b)` (or `cache(x,A,b)`). Both in-place and out of place work everywhere. Updating of the factorization or KSP solver's cache is done automatically using `cache.isfresh` flag.
2. With how things are structured, for every new `SciMLLinearSolveAlgorithm`, you only need to define two functions: `init_cacheval(alg::NewAlg, A, b, u)`, and `SciMLBase.solve(cache, alg)` which calls `init_cacheval()` when `cache.isfresh == true`. Check out `src/factorization.jl` for examples.
3. Check out `DefaultLinSolve` in `src/default.jl`
4. Wrappers for `Krylov.jl`, and `IterativeSolvers.jl` are largely done. I've added tests with different keyword arguments.

**Comment by vpuri3** [comment]
TODO:
1. add more complicated tests
2. write out `ComposedPreconditioner`, `ScaleVector` as in `diffeqbase/linear_nonlienar`, and store preconditioner in `cache.Pl = cache.Pl * alg.Pl`, `cache.Pr = alg.Pr * cache.Pr`. Current wrapper implementation just uses `alg.Pr,Pr`.

**Comment by ChrisRackauckas** [comment]
```
TODO:

add more complicated tests
write out ComposedPreconditioner, ScaleVector as in diffeqbase/linear_nonlienar, and store preconditioner in cache.Pl = cache.Pl * alg.Pl, cache.Pr = alg.Pr * cache.Pr. Current wrapper implementation just uses alg.Pr,Pr.
inject callback(iterator) in IterativeSolvers.jl wrapper
OrdinaryDiffEq integration?
```

definitely not in this PR.

**Comment by ChrisRackauckas** [comment]
Retriggering the tests for https://github.com/SciML/SciMLBase.jl/pull/119

File: src/default.jl [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    if alg.isset
+      linalg = if A isa Matrix
+          if ArrayInterface.can_setindex(x) && (size(A,1) <= 100 ||
+                                                (p.openblas && size(A,1) <= 500)
+                                               )
+              DefaultFactorization(;fact_alg=:(RecursiveFactorization.lu!))
+          else
+              LUFactorization()
+          end
+      elseif A isa Union{Tridiagonal,} # ForwardSensitivityJacobian
```
Comment (ChrisRackauckas) [inline src/default.jl:33]: Is this not a syntax error?

File: src/default.jl [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+      elseif !(A isa AbstractDiffEqOperator)
+          QRFactorization()
+      else
+          IterativeSolversJL_GMRES()
+      end
+
+      @set! alg.linalg = linalg
+    end
+
+    SciMLBase.solve(cache, alg.linalg, args...; kwargs...)
```
Comment (ChrisRackauckas) [inline src/default.jl:50]: @YingboMa do you know if hoisting this into the if statements will avoid a dynamic dispatch?

File: src/wrappers.jl [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+  (Krylov.bilq!       => Krylov.BilqSolver          ),
+  (Krylov.minres_qlp! => Krylov.MinresQlpSolver     ),
+  (Krylov.qmr!        => Krylov.QmrSolver           ),
+  (Krylov.gmres!      => Krylov.GmresSolver         ),
+  (Krylov.fom!        => Krylov.FomSolver           ),
+ )
+
+function init_cacheval(alg::KrylovJL, A, b, u)
+
+    KS = KrylovJL_solvers[alg.KrylovAlg]
```
Comment (ChrisRackauckas) [inline src/wrappers.jl:103]: This is going to be unstable. Mostly worried about the compile time effect. It's better to make that a function with a bunch of dispatches.

**Comment by ChrisRackauckas** [comment]
Let's cut this here. Default preconditioners can be a follow up PR. @YingboMa can you give this a review and see what you would recommend?

**Comment by ChrisRackauckas** [comment]
Tests are failing. Seems like a tolerance thing?

File: test/runtests.jl [hunk 4]
```diff
@@ -52,33 +56,47 @@ using Test
...
+#                    :cholesky, :cholesky!,
+#   #                :ldlt, :ldlt!,
+#                    :bunchkaufman, :bunchkaufman!,
+#                    :lq, :lq!,
+#                    :svd, :svd!,
+#                    :(LinearAlgebra.factorize), 
+#                   )
+#       kwargs = :(fact_alg=$fact_alg,)
+#       test_interface(alg, kwargs, prob1, prob2, prob3)
+#   end
```
Comment (ChrisRackauckas) [inline test/runtests.jl:77]: Never comment out tests 😱

Comment (ChrisRackauckas) [reply test/runtests.jl:77]: Do these just not apply anymore? If so, remove them if the same pieces are tested elsewhere.

File: src/default.jl (hunk 1 above)
Comment (ChrisRackauckas) [reply src/default.jl:34]: remove the `,` anyways.

Comment (ChrisRackauckas) [reply src/default.jl:34]: Well, also, no union.

File: src/default.jl [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+          if ArrayInterface.can_setindex(x) && (size(A,1) <= 100 ||
+                                                (p.openblas && size(A,1) <= 500)
+                                               )
+              DefaultFactorization(;fact_alg=:(RecursiveFactorization.lu!))
+          else
+              LUFactorization()
+          end
+      elseif A isa Union{Tridiagonal,} # ForwardSensitivityJacobian
+          DefaultFactorization(;fact_alg=lu!)
+      elseif A isa Union{SymTridiagonal}
```
Comment (ChrisRackauckas) [inline src/default.jl:36]: no union

Comment (vpuri3) [reply src/default.jl]: done

File: test/runtests.jl (hunk 4 above)
Comment (vpuri3) [reply test/runtests.jl]: i wrote them messily. was commenting to find erroring alg. using testsets instead

File: src/default.jl (hunk 1 above)
Comment (vpuri3) [reply src/default.jl]: done

**Comment by vpuri3** [comment]
seems like there's a problem with `cholesky!` factorization. `cholesky` works fine though. can you retrigger CI?

**Comment by ChrisRackauckas** [comment]
You can always retrigger CI by closing and reopening the PR BTW.

File: Project.toml [hunk 6]
```diff
@@ -5,11 +5,15 @@ version = "0.1.0"
 
 [deps]
 ArrayInterface = "4fba245c-0d91-5ea0-9b3e-6abc04ee57a9"
+IterativeSolvers = "42fd0dbc-a981-5370-80f2-aaf504508153"
```
Comment (YingboMa) [inline Project.toml:8]: Compats

File: src/common.jl [hunk 7]
```diff
@@ -1,80 +1,126 @@
...
+    u::Tu
     p::Tp
     alg::Talg
-    cacheval::Tc
-    isfresh::Bool
-    Pr::Tr
-    Pl::Tl
+    cacheval::Tc  # store alg cache here 
+    isfresh::Bool # false => cacheval is set wrt A, true => update cacheval wrt A
+#
```
Comment (YingboMa) [inline src/common.jl:9]:
```suggestion
```

File: src/common.jl [hunk 8]
```diff
@@ -1,80 +1,126 @@
...
+    Pl::Tl        # store final preconditioner here. not being used rn
+    Pr::Tr        # wrappers are using preconditioner in cache.alg for now
 end
 
-function set_A(cache, A)
+function set_A(cache::LinearCache, A) # and ! to function name
```
Comment (YingboMa) [inline src/common.jl:14]:
```suggestion
function set_A(cache::LinearCache, A)
```
We don't add `!` to non-mutating function.

File: src/default.jl (hunk 2 above)
Comment (YingboMa) [reply src/default.jl:51]: Yeah, moving the solve into the if statements should be better.

File: src/common.jl [hunk 9]
```diff
@@ -1,80 +1,126 @@
...
+                        kwargs...,
+                       )
+    @unpack A, b, u0, p = prob
 
-    A = alias_A ? A : copy(A)
-    b = alias_b ? b : copy(b)
+    u0 = (u0 == nothing) ? zero(b) : u0
+
+    cacheval = init_cacheval(alg, A, b, u0)
+    isfresh = cacheval == nothing
```
Comment (YingboMa) [inline src/common.jl:55]:
```suggestion
    isfresh = cacheval === nothing
```

File: src/common.jl [hunk 10]
```diff
@@ -1,80 +1,126 @@
...
+
+function SciMLBase.init(prob::LinearProblem, alg, args...;
+                        alias_A = false, alias_b = false,
+                        kwargs...,
+                       )
+    @unpack A, b, u0, p = prob
 
-    A = alias_A ? A : copy(A)
-    b = alias_b ? b : copy(b)
+    u0 = (u0 == nothing) ? zero(b) : u0
```
Comment (YingboMa) [inline src/common.jl:52]:
```suggestion
    u0 = (u0 === nothing) ? zero(b) : u0
```

File: src/wrappers.jl (hunk 3 above)
Comment (YingboMa) [reply src/wrappers.jl:103]: Agreed.

File: Project.toml (hunk 6 above)
Comment (ChrisRackauckas) [reply Project.toml:8]: Still missing.

Comment (vpuri3) [reply Project.toml:8]: yep sorry was in b2b meetings. adding `IterativeSolvers = "0.9.2"`.

Comment (vpuri3) [reply Project.toml:8]: done

File: src/wrappers.jl (hunk 3 above)
Comment (vpuri3) [reply src/wrappers.jl]: changed it to a function with if statements. have a look now?

Comment (ChrisRackauckas) [reply src/wrappers.jl]: that should be fine. I would've done dispatch, but both will evaluate at compile time.

File: src/default.jl (hunk 2 above)
Comment (vpuri3) [reply src/default.jl:51]: the if block only runs once to select the linear algorithm

**Comment by vpuri3** [comment]
the tests are failing because `cholesky!` is modifying the `A` matrix which is being used in tests  as `@test $A * $x \approx $b`. fixing this

File: src/common.jl [hunk 11]
```diff
@@ -1,80 +1,125 @@
...
-    else
-        fact = nothing
-        Tfact = Any
-    end
-    Pr = nothing
-    Pl = nothing
+init_cacheval(alg::SciMLLinearSolveAlgorithm, A, b, u) = nothing
+
+function SciMLBase.init(prob::LinearProblem, alg, args...;
+                        alias_A = false, alias_b = false,
```
Comment (ChrisRackauckas) [inline src/common.jl:46]: these arguments should be handling the aliasing

**Comment by vpuri3** [comment]
here's the issue!
```julia
julia> using LinearAlgebra; n=8; A=Matrix(I,n,n)/2; fa=cholesky!(A)
Cholesky{Float64, Matrix{Float64}}
U factor:
8×8 UpperTriangular{Float64, Matrix{Float64}}:
 0.707107  0.0       0.0       0.0       0.0       0.0       0.0       0.0
  ⋅        0.707107  0.0       0.0       0.0       0.0       0.0       0.0
  ⋅         ⋅        0.707107  0.0       0.0       0.0       0.0       0.0
  ⋅         ⋅         ⋅        0.707107  0.0       0.0       0.0       0.0
  ⋅         ⋅         ⋅         ⋅        0.707107  0.0       0.0       0.0
  ⋅         ⋅         ⋅         ⋅         ⋅        0.707107  0.0       0.0
  ⋅         ⋅         ⋅         ⋅         ⋅         ⋅        0.707107  0.0
  ⋅         ⋅         ⋅         ⋅         ⋅         ⋅         ⋅        0.707107

julia> fa
Cholesky{Float64, Matrix{Float64}}
U factor:
8×8 UpperTriangular{Float64, Matrix{Float64}}:
 0.707107  0.0       0.0       0.0       0.0       0.0       0.0       0.0
  ⋅        0.707107  0.0       0.0       0.0       0.0       0.0       0.0
  ⋅         ⋅        0.707107  0.0       0.0       0.0       0.0       0.0
  ⋅         ⋅         ⋅        0.707107  0.0       0.0       0.0       0.0
  ⋅         ⋅         ⋅         ⋅        0.707107  0.0       0.0       0.0
  ⋅         ⋅         ⋅         ⋅         ⋅        0.707107  0.0       0.0
  ⋅         ⋅         ⋅         ⋅         ⋅         ⋅        0.707107  0.0
  ⋅         ⋅         ⋅         ⋅         ⋅         ⋅         ⋅        0.707107

julia> A
8×8 Matrix{Float64}:
 0.707107  0.0       0.0       0.0       0.0       0.0       0.0       0.0
 0.0       0.707107  0.0       0.0       0.0       0.0       0.0       0.0
 0.0       0.0       0.707107  0.0       0.0       0.0       0.0       0.0
 0.0       0.0       0.0       0.707107  0.0       0.0       0.0       0.0
 0.0       0.0       0.0       0.0       0.707107  0.0       0.0       0.0
 0.0       0.0       0.0       0.0       0.0       0.707107  0.0       0.0
 0.0       0.0       0.0       0.0       0.0       0.0       0.707107  0.0
 0.0       0.0       0.0       0.0       0.0       0.0       0.0       0.707107
```

File: src/common.jl (hunk 11 above)
Comment (vpuri3) [reply src/common.jl:46]: line 60-61 of the same file. 

```julia
    A = alias_A ? A : deepcopy(A)
    b = alias_b ? b : deepcopy(b)
```
This was in `common.jl` when I started the file.

**Comment by ChrisRackauckas** [comment]
No, the entire purpose of the mutating cholesky decomposition is that... it's mutating. Are you mutating the copied one?

**Comment by vpuri3** [comment]
ok so when we do `fa = cholesky!(A)`, the space in `A` is where the factors are stored. So we can't just do `A * x` and the tests are written incorrectly.

**Comment by ChrisRackauckas** [comment]
I am digging in and noticed that the reason is that the API was completely bungled. Making the cache and algs callable are not only not part of the API, but it's also a really bad idea and the reason for the failures. This is all really far off so I'm taking this PR over to put it back on the right track.

**Comment by vpuri3** [comment]
I was attempting to mimic the api in `diffeqbase/linear_nonlinear` where you could do `DefaultLinSolve()(x, A, b)`, where DefaultLinSolve is an algorithm, and the calls to set_A, set_b, and set_u would just be within `alg()(x,A,b)`.

**Comment by ChrisRackauckas** [comment]
The design is not supposed to mimic that. It's supposed to be matching the common interface design. 

https://scimlbase.sciml.ai/dev/#Common-Interface-High-Level

The whole purpose is to move linear solvers away from a hacky side design to something that follows the same flow as nonlinear solvers, diffeq solvers, optimization solvers, etc. so that the whole universe of numerical problems follows the same exact interface.

**Comment by vpuri3** [comment]
i agree with the point on api, and apologies for messing with the API without discussing.

based on last commit, the way to do things is:
```julia
cache = set_A(cache, A)
cache = set_b(cache, b)
cache = set_u(cache, u)
solve(cache)
```

would it make sense to allow for this?
```julia
function solve(prob, cache)
    cache = set_A(cache, prob.A)
    cache = set_b(cache, prob.b)
    cache = set_u(cache, prob.u0)
    return solve(cache)
end

prob = LinearProblem(A,b;u0=u)
solve(prob,cache)
```

**Comment by ChrisRackauckas** [comment]
I don't understand the use case. What is it?

**Pull request merged**